numpy/lib/function_base.py

```diff
@@ -1320,6 +1320,258 @@ def gradient(f, *varargs, **kwargs):
+    one-sides (forward or backwards) differences at the boundaries. The
+    returned gradient hence has the same shape as the input array.
+    
+    Parameters
+    ----------
+    v : list of numpy arrays
+        Each of these array is the N-dimensional projection 
+        of the vector field on the corresponding axis. 
+    
+    varargs : scalar or list of scalar, optional
+        N scalars specifying the sample distances for each dimension,
+        i.e. `dx`, `dy`, `dz`, ... Default distance: 1.
+        single scalar specifies sample distance for all dimensions.
+        if `axis` is given, the number of varargs must equal the number of axes.
+    edge_order : {1, 2}, optional
+        Gradient is calculated using N\ :sup:`th` order accurate differences
+        at the boundaries. Default: 1.
+        
+    axis : None or int or tuple of ints, optional
+        Gradient is calculated only along the given axis or axes
+        The default (axis = None) is to calculate the gradient for 
+        all the axes of the input array.
+        axis may be negative, in which case it counts from the last to the first axis.
+        
+    Returns
+    -------
+    curl : list of numpy arrays
+        The output corresponds to the curl of the input 3-D vector field.
+        This means that the output has the form 
+        (dAz/dy-dAy/dz, dAx/dz-dAz/dx, dAy/dx-dAx/dy) of an input vector 
+        field A = (Ax, Ay, Ax).
+    
+    Example
+    -------
+    The following example shows in matplotlib how the vector field (0,0,x*y) has the 
+    correct value, nonzero only in the third component of the resulting vector field
+    after applying curl
+    
+    >>> import matplotlib.pylab as plt
+    >>> import numpy as np
+    >>> X,Y,Z=np.mgrid[0:200,0:200,0:200]
+    >>> a0=X*Y
+    >>> a1=np.zeros(np.shape(X))
+    >>> a2=np.zeros(np.shape(X))
+    >>> a=[a0,a1,a2]
+    >>> [cx,cy,cz]=np.curl(a)
+    >>> plt.imshow(cx[:,:,1])
+    >>> plt.colorbar()
+    >>> plt.show()
+    >>> plt.imshow(cy[:,:,1])
+    >>> plt.colorbar()
+    >>> plt.show()
+    >>> plt.imshow(cz[:,:,1])
+    >>> plt.colorbar()
+    >>> plt.show()
+    """
+    N = [0]*len(v)
+    for i, v_c in enumerate(v):
+        v_c = np.asanyarray(v_c)
+	N[i] = len(v_c.shape) 
+        # Extract the number of dimensions from every component v_c
+    if False in [N[i] == 3 for i in range(len(v))]: 
+        #Check if all components are the same size
+        raise ValueError("Not all components of input vector field"
+                         " have three dimensions")
+    else:
+        if False in [np.shape(v[0]) == np.shape(v[i]) for i in range(len(v))]:
+            raise ValueError("Not all components of input are the same size")
+        else:
+            N = N[0] 
+        # If all vector field components are the same shape,
+        #N becomes the number of dimensions
+
+    outvals=[np.zeros(np.shape(v_c))] * 3 #Initialize output vector field
```

Comment:
Hello everyone,
Thank you for all your suggestions, they will definitely improve the functions. I will start implementing them, I guess it will take me a couple of days to review everything and make some profiles to filter the most efficient ones. By the way, in my last commit I force edge_order to 0 when calling gradient because if it is 1 something wrong happens in the edges, and the examples don't return arrays of zeros (in the edges there are 0.5 and -0.5). With edge_order = 0 it does not happen.
Thanks again